core/src/main/java/org/springframework/security/authorization/method/MethodAccessDeniedHandlerResolver.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2002-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.authorization.method;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.context.ApplicationContext;
+import org.springframework.core.log.LogMessage;
+import org.springframework.util.ClassUtils;
+
+public final class MethodAccessDeniedHandlerResolver {
+
+	private final Log logger = LogFactory.getLog(getClass());
+
+	private final Map<Class<? extends MethodAccessDeniedHandler>, MethodAccessDeniedHandler> handlerCache = new ConcurrentHashMap<>();
+
+	private ApplicationContext applicationContext;
+
+	public MethodAccessDeniedHandler resolve(Method method) {
+		DeniedHandler deniedHandler = findDeniedHandler(method);
+		// FIXME should we look into the ApplicationContext for a bean of type?
+		return (deniedHandler != null) ? getBeanOrNew(deniedHandler.value()) : null;
+	}
+
+	private DeniedHandler findDeniedHandler(Method method) {
+		DeniedHandler methodAnnotation = AuthorizationAnnotationUtils.findUniqueAnnotation(method, DeniedHandler.class);
+		if (methodAnnotation != null) {
+			return methodAnnotation;
+		}
+		return AuthorizationAnnotationUtils.findUniqueAnnotation(method.getDeclaringClass(), DeniedHandler.class);
+	}
+
+	private MethodAccessDeniedHandler getBeanOrNew(Class<? extends MethodAccessDeniedHandler> handlerClass) {
+		if (this.applicationContext != null && this.applicationContext.getBeanNamesForType(handlerClass).length == 1) {
```

Comment:
I'm not so sure about inferring something like this. It makes it a little tricky if I want some methods to use a managed bean when others don't, especially when those are in same the type hierarchy.

For example, if I have something like this:

```java
public class MyAccessDeniedHandler implements MethodAccessDeniedHandler {
    // ...
}
```

that is not a bean, and then

```java
@Component 
public class MyComponentAccessDeniedHandler extends MyAccessDeniedHandler {
    // ...
]
```

then when doing the following:

```java
@DeniedHandler(MyAccessDeniedHandler.class)
```

I believe this resolver could select the incorrect handler.

Instead, I think resolving by bean could either wait for a future enhancement, or perhaps the annotation could instead have two properties, say `handlerClass` and `handlerBeanName`, in order to make the intent clear.

Reply:
While I understand the concern, I worry about deviating from the semantics of how `Autowired` behaves. A configuration of `@Autowired MyAccessDeniedHandler` would also resolve to `MyComponentAcccessDeniedHandler`. I think changing these semantics is a bit surprising to users. Differences like this would make it more difficult to use.

A benefit of this approach is that if `@DeniedHandler` defaults to `MethodAccessDeniedHandler` then a default Bean of `NullMethodAccessDeniedHandler` could be provided. Users wanting to override that default could provide their own `MethodAccessDeniedHandler` bean.

Reply:
Agreed, my point is mainly that the two approaches are incompatible and that we should choose one or the other. I agree that bean-based is the most reasonable default behavior.